test/src/unit-capi-subarray_partitioner-dense.cc

```diff
@@ -0,0 +1,2095 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests the `capi subarray_partitioner` class for dense arrays.
+ */
+
+#include "test/src/helpers.h"
+#include "test/src/vfs_helpers.h"
+#include "tiledb/sm/c_api/tiledb_struct_def.h"
+#include "tiledb/sm/subarray/subarray_partitioner.h"
+
+#ifdef _WIN32
+#include "tiledb/sm/filesystem/win.h"
+#else
+#include "tiledb/sm/filesystem/posix.h"
+#endif
+
+#include <catch.hpp>
+#include <iostream>
+
+using namespace tiledb::common;
+using namespace tiledb::sm;
+using namespace tiledb::test;
+
+/* ********************************* */
+/*         STRUCT DEFINITION         */
+/* ********************************* */
+
+struct CAPISubarrayPartitionerDenseFx {
+  tiledb_ctx_t* ctx_;
+  tiledb_vfs_t* vfs_;
+  const std::vector<std::unique_ptr<SupportedFs>> fs_vec_;
+  std::string temp_dir_;
+  std::string array_name_;
+  const char* ARRAY_NAME = "subarray_partitioner_dense";
+  tiledb_array_t* array_ = nullptr;
+  uint64_t memory_budget_ = 1024 * 1024 * 1024;
+  uint64_t memory_budget_var_ = 1024 * 1024 * 1024;
+
+  CAPISubarrayPartitionerDenseFx();
+  ~CAPISubarrayPartitionerDenseFx();
+
+  /**
+   * Helper function that creates a default 1D array, with the
+   * input tile and cell order.
+   */
+  void create_default_1d_array(
+      tiledb_layout_t tile_order, tiledb_layout_t cell_order);
+
+  /**
+   * Helper function that creates a default 2D array, with the
+   * input tile and cell order.
+   */
+  void create_default_2d_array(
+      tiledb_layout_t tile_order, tiledb_layout_t cell_order);
+
+  /** Helper function that writes to the default 1D array. */
+  void write_default_1d_array();
+
+  /** Helper function that writes to the default 2D array. */
+  void write_default_2d_array();
+
+  /**
+   * Helper function to test the subarray partitioner for the given arguments.
+   */
+  template <class T>
+  void test_subarray_partitioner(
+      Layout subarray_layout,
+      const SubarrayRanges<T>& ranges,
+      const std::vector<SubarrayRanges<T>>& partitions,
+      const std::string& attr,  // Attribute to set the budget for
+      uint64_t budget,
+      bool unsplittable = false);
+
+  /**
+   * Helper function to test the subarray partitioner for the given arguments.
+   * This is different from the above in that it tests the memory
+   * budget.
+   */
+  template <class T>
+  void test_subarray_partitioner(
+      Layout subarray_layout,
+      const SubarrayRanges<T>& ranges,
+      const std::vector<SubarrayRanges<T>>& partitions,
+      uint64_t budget,
```

Comment:
budget and budget_var are the names in use in the ancestor source modules, 
unit-SubarrayPartitioner-dense.cc and unit-SubarrayPartitioner-Sparse.cc 
from which 
unit-capi-SubarrayPartioner-dense.cc and unit-capi-SubarrayPartitioner-sparse.cc were derived.
Since functionality is expected to be added, I decline to rename at this time until all to avoid introducing noise for comparing/syncing